File: backend/jobs.py
"""Live Jobs API — reads the bin/job registry and streams progress to the SPA.

Layer 2 of the Live Jobs design (docs/superpowers/specs/2026-06-30-workspace-live-jobs-design.md).

The registry is a directory of atomic JSON files written by `bin/job`
(`$WORKSPACE/tmp/jobs/<id>.json`). This router only READS them — it never
writes, so it stays fully decoupled from producers.

  GET /api/jobs          -> {"jobs": [ …records… ]}  (running first, newest first)
  GET /api/jobs/stream   -> text/event-stream; snapshot on connect, then a framed
                            record list whenever it changes. Source of truth is
                            task_registry (fed by task_ingest's scan loop); the
                            idle keepalive tick re-applies the 60s terminal
                            window so finished cards age out without an event.

Fail-soft everywhere: a malformed/partial file is skipped, never fatal.
"""

from __future__ import annotations

import asyncio
import json
import re
import time
from pathlib import Path

from fastapi import APIRouter
from fastapi.responses import (HTMLResponse, PlainTextResponse,
                               StreamingResponse)

from . import task_registry
from .vault_store import WORKSPACE

_ID_RE = re.compile(r"^[A-Za-z0-9_]+$")   # job ids are safe tokens; reject the rest

router = APIRouter()

JOBS_DIR = WORKSPACE / "tmp" / "jobs"

RETAIN_SECS = 60          # terminal jobs older than this are dropped from output
                          # (tighter than the registry's own RETAIN_TERMINAL_S —
                          # /api/jobs has always aged terminal rows out faster)

_KEEPALIVE_S = 15.0       # idle SSE tick: keepalive comment + terminal-window re-check

# Internal bookkeeping fields we don't leak to the client.
_PRIVATE = ("_updatedEpoch", "_pctExplicit")


def _native(rec: dict) -> dict:
    """Reconstruct the bin/job native record from a registry entry: the
    ingested payload verbatim, private fields stripped, `stalled` injected
    when the registry derived it. This is the compat contract — the overlay
    and /jobs/live were built against these exact fields."""
    out = dict((rec.get("extra") or {}).get("native") or {})
    for k in _PRIVATE:
        out.pop(k, None)
    if rec.get("state") == "stalled":
        updated = (rec.get("extra") or {}).get("updated_epoch") or 0
        out["stalled"] = int(time.time() - updated) if updated else 1
    return out


def _read_all() -> list[dict]:
    """All current job records from the registry (source=job only), cleaned +
    sorted running-first / newest-first — the same output contract the
    directory-globbing implementation had. Re-applies the legacy RETAIN_SECS
    terminal window on top of the registry's own (longer) RETAIN_TERMINAL_S,
    using the registry record's own `updated` timestamp — not the native
    payload's — since that's what the registry actually ages records by.
    """
    now_ms = time.time() * 1000
    recs: list[dict] = []
    for rec in task_registry.list_tasks(source="job"):
        native = _native(rec)
        if not native.get("id"):
            continue
        if (native.get("status") in ("done", "failed")
                and now_ms - rec["updated"] > RETAIN_SECS * 1000):
            continue
        recs.append(native)

    order = {"running": 0, "failed": 1, "done": 2}
    recs.sort(key=lambda r: (order.get(r.get("status"), 3),
                             _neg(r.get("startedAt", ""))))
    return recs


def _neg(s: str):
    # sort strings descending by negating their sort position via reverse tuple
    return tuple(-ord(c) for c in s)


def _sse(obj) -> str:
    return f"data: {json.dumps(obj, separators=(',', ':'))}\n\n"


@router.get("/api/jobs")
async def jobs():
    return {"jobs": _read_all()}


async def _stream_gen():
    queue = task_registry.subscribe()
    try:
        # emit an immediate snapshot so the client renders without waiting a tick
        snap = _read_all()
        last = json.dumps(snap, separators=(",", ":"))
        yield _sse({"jobs": snap})
        while True:
            try:
                rec = await asyncio.wait_for(queue.get(), timeout=_KEEPALIVE_S)
            except asyncio.TimeoutError:
                if not task_registry.is_subscribed(queue):
                    return          # dropped (QueueFull): end the stream; client resnapshots
                # This tick doubles as the self-heal pass for the time-based
                # RETAIN_SECS window: a terminal job crossing the 60s cutoff
                # produces no registry event, so re-check the list here.
                cur = _read_all()
                key = json.dumps(cur, separators=(",", ":"))
                if key != last:
                    last = key
                    yield _sse({"jobs": cur})
                else:
                    yield ": keepalive\n\n"
                continue
            if rec.get("source") != "job":
                continue
            cur = _read_all()
            key = json.dumps(cur, separators=(",", ":"))
            if key != last:
                last = key
                yield _sse({"jobs": cur})
    finally:
        task_registry.unsubscribe(queue)


@router.get("/api/jobs/stream")
async def jobs_stream():
    return StreamingResponse(_stream_gen(), media_type="text/event-stream",
                             headers={"Cache-Control": "no-cache",
                                      "X-Accel-Buffering": "no"})


# ---------------------------------------------------------------------------
# Terminal-mirror band-aid: a per-job raw log tail + a standalone page that
# renders it. Deliberately OUTSIDE the SPA bundle so it works on devices whose
# cached PWA shell predates the Live Jobs overlay (the mobile "not showing"
# case). `bin/job-run` tees the child's stdout/stderr to tmp/jobs/<id>.log.
# ---------------------------------------------------------------------------

def _tail(path: Path, n: int) -> str:
    try:
        data = path.read_text(errors="replace")
    except Exception:
        return ""
    lines = data.splitlines()
    return "\n".join(lines[-n:])


@router.get("/api/jobs/{jid}/log", response_class=PlainTextResponse)
async def job_log(jid: str, tail: int = 300):
    """Last `tail` lines of a job's mirrored terminal output (plain text)."""
    if not _ID_RE.match(jid):
        return PlainTextResponse("", status_code=404)
    tail = max(1, min(tail, 2000))
    body = _tail(JOBS_DIR / f"{jid}.log", tail)
    return PlainTextResponse(body, headers={"Cache-Control": "no-store"})


_LIVE_HTML = """<!doctype html><html lang="en"><head>
<meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="color-scheme" content="dark light"><title>Live Jobs</title>
<style>
:root{--bg:#0d0f13;--panel:#171a20;--bd:#2a2f38;--fg:#e8e8ea;--mut:#9aa0aa;
  --accent:#5b9dff;--ok:#3fbf6f;--err:#e5534b;--amber:#d8a24a}
*{box-sizing:border-box}
body{margin:0;background:var(--bg);color:var(--fg);
  font:14px/1.45 system-ui,-apple-system,Segoe UI,sans-serif;
  padding:14px 12px calc(24px + env(safe-area-inset-bottom))}
h1{font-size:15px;margin:2px 4px 12px;display:flex;align-items:center;gap:8px}
h1 .dot{width:8px;height:8px;border-radius:50%;background:var(--mut)}
h1.live .dot{background:var(--accent);box-shadow:0 0 0 0 var(--accent);animation:p 1.6s infinite}
@keyframes p{0%{box-shadow:0 0 0 0 rgba(91,157,255,.5)}70%{box-shadow:0 0 0 7px rgba(91,157,255,0)}100%{box-shadow:0 0 0 0 rgba(91,157,255,0)}}
.job{background:var(--panel);border:1px solid var(--bd);border-left:3px solid var(--bd);
  border-radius:11px;padding:11px 12px;margin:0 0 11px}
.job.running{border-left-color:var(--accent)}
.job.done{border-left-color:var(--ok)}
.job.failed{border-left-color:var(--err)}
.job.stalled{border-left-color:var(--amber)}
.top{display:flex;align-items:baseline;gap:8px;margin-bottom:7px}
.lbl{font-weight:600;flex:1;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}
.pct{font-variant-numeric:tabular-nums;color:var(--mut)}
.track{height:6px;border-radius:3px;background:#2a2f38;overflow:hidden;margin-bottom:7px}
.fill{height:100%;width:0;background:var(--accent);border-radius:3px;transition:width .4s ease}
.job.done .fill{background:var(--ok)}.job.failed .fill{background:var(--err)}
.meta{color:var(--mut);font-size:12px;margin-bottom:8px}
pre{margin:0;max-height:44vh;overflow:auto;background:#0a0c10;border:1px solid var(--bd);
  border-radius:8px;padding:9px 10px;font:12px/1.4 ui-monospace,Menlo,Consolas,monospace;
  white-space:pre-wrap;word-break:break-word;color:#cfd3da}
.empty{color:var(--mut);text-align:center;padding:40px 0}
.err{color:#e5847e;font-size:12px;margin-bottom:6px}
</style></head><body>
<h1 id="h"><span class="dot"></span><span id="ht">Live Jobs</span></h1>
<div id="root"><div class="empty">Loading…</div></div>
<script>
const R=document.getElementById('root'),H=document.getElementById('h'),HT=document.getElementById('ht');
const stick={};              // per-job: is the log scrolled to bottom?
const esc=s=>String(s==null?'':s).replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;');
async function logFor(id){try{const r=await fetch('/api/jobs/'+id+'/log?tail=400',{cache:'no-store'});
  return r.ok?await r.text():'';}catch(_){return'';}}
async function tick(){
  let jobs=[];try{const r=await fetch('/api/jobs',{cache:'no-store'});jobs=(await r.json()).jobs||[];}
  catch(_){HT.textContent='Live Jobs — offline';H.classList.remove('live');return;}
  const running=jobs.filter(j=>j.status==='running').length;
  H.classList.toggle('live',running>0);
  HT.textContent=running>0?running+' running':(jobs.length?'Jobs done':'No jobs');
  if(!jobs.length){R.innerHTML='<div class="empty">No jobs running.</div>';return;}
  const logs=await Promise.all(jobs.map(j=>logFor(j.id)));
  R.innerHTML=jobs.map((j,i)=>{
    const cls=['job',j.status];if(j.stalled)cls.push('stalled');
    const pct=j.status==='done'?100:(j.pct!=null?j.pct:0);
    const pl=j.status==='done'?'✓':j.status==='failed'?'✕':(j.pct!=null?Math.round(j.pct)+'%':'');
    const meta=[j.detail,j.eta?('ETA '+j.eta+'s'):'',j.stalled?('no update '+j.stalled+'s'):'']
      .filter(Boolean).map(esc).join(' · ');
    const log=logs[i]||'';
    return `<div class="${cls.join(' ')}">
      <div class="top"><span class="lbl">${esc(j.label||j.id)}</span><span class="pct">${pl}</span></div>
      <div class="track"><div class="fill" style="width:${pct}%"></div></div>
      ${meta?`<div class="meta">${meta}</div>`:''}
      ${j.error?`<div class="err">${esc(j.error)}</div>`:''}
      ${log?`<pre data-id="${esc(j.id)}">${esc(log)}</pre>`:''}
    </div>`;}).join('');
  R.querySelectorAll('pre').forEach(p=>{const id=p.dataset.id;
    if(stick[id]!==false)p.scrollTop=p.scrollHeight;   // auto-follow unless user scrolled up
    p.onscroll=()=>{stick[id]=p.scrollHeight-p.scrollTop-p.clientHeight<24;};});
}
tick();setInterval(tick,1500);
</script></body></html>"""


@router.get("/jobs/live", response_class=HTMLResponse)
async def jobs_live():
    """Standalone raw-tail page — SW never caches this route, so it renders on
    devices whose PWA shell is stale. Poll-based; no SPA dependency."""
    return HTMLResponse(_LIVE_HTML, headers={"Cache-Control": "no-store"})

File: frontend-overrides/js/redesign/live/jobs.js
// Live Jobs overlay — Layer 3 of the Live Jobs design.
// docs/superpowers/specs/2026-06-30-workspace-live-jobs-design.md
//
// A self-contained progress overlay: it injects its own CSS + container onto
// <body>, opens /api/jobs/stream (SSE), keeps a jobs map, and renders live
// progress bars imperatively. It is deliberately DECOUPLED from the SPA's
// render() cycle so a chat re-render can never destroy or duplicate it.
//
// UX:
//   * A stack of slim bars, bottom-right, above the composer. Visible only when
//     >=1 job exists; auto-collapses ~4s after the last job finishes.
//   * A header badge "N running" toggles the stack collapsed/expanded.
//   * Bars for the currently-open chat thread get an accent rail; jobs from
//     other threads are dimmed but still listed (so Frank always sees every
//     process, which was the whole point).
//   * done -> green check, failed -> red + error, stalled -> amber "no update".
//
// Fail-soft: SSE errors reconnect with backoff; a bad payload is ignored; if the
// endpoint is missing the overlay simply stays empty and invisible.

import { runtime } from './runtime.js';

const STREAM = '/api/jobs/stream';
const FALLBACK_POLL = '/api/jobs';
const FADE_AFTER_MS = 4000;      // collapse the panel this long after it empties

let els = null;                  // { root, badge, list }
let jobs = [];                   // latest record list from the stream
let es = null;
let backoff = 1000;
let emptySince = 0;
let pollTimer = null;

// ---- formatting -----------------------------------------------------------

function fmtBytes(n) {
  if (n == null) return '';
  const u = ['B', 'KB', 'MB', 'GB', 'TB'];
  let i = 0, v = n;
  while (v >= 1024 && i < u.length - 1) { v /= 1024; i++; }
  return `${v >= 100 || i === 0 ? Math.round(v) : v.toFixed(1)} ${u[i]}`;
}

function fmtRate(bps) {
  if (!bps || bps <= 0) return '';
  return `${fmtBytes(bps)}/s`;
}

function fmtEta(s) {
  if (s == null || s < 0) return '';
  s = Math.round(s);
  if (s < 60) return `${s}s`;
  const m = Math.floor(s / 60), ss = s % 60;
  if (m < 60) return `${m}m ${String(ss).padStart(2, '0')}s`;
  const h = Math.floor(m / 60);
  return `${h}h ${String(m % 60).padStart(2, '0')}m`;
}

function esc(s) {
  return String(s == null ? '' : s)
    .replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
}

function kindIcon(kind) {
  switch (kind) {
    case 'download': return '⬇';
    case 'upload': return '⬆';
    case 'verify': return '✓';
    case 'render': return '🎬';
    case 'build': return '🔧';
    default: return '⏳';
  }
}

// ---- DOM setup ------------------------------------------------------------

const CSS = `
#live-jobs{position:fixed;right:16px;bottom:84px;z-index:6000;width:340px;max-width:calc(100vw - 32px);
  display:none;flex-direction:column;gap:6px;font:13px/1.35 var(--font,system-ui,sans-serif);pointer-events:none}
#live-jobs.show{display:flex}
#live-jobs .lj-badge{align-self:flex-end;pointer-events:auto;cursor:pointer;user-select:none;
  display:inline-flex;align-items:center;gap:7px;padding:5px 11px;border-radius:999px;
  background:var(--panel,rgba(30,30,34,.92));border:1px solid var(--border,rgba(127,127,127,.28));
  color:var(--text,#e8e8ea);box-shadow:0 4px 16px rgba(0,0,0,.28);backdrop-filter:blur(8px)}
#live-jobs .lj-badge .lj-dot{width:7px;height:7px;border-radius:50%;background:var(--accent,#5b9dff);
  box-shadow:0 0 0 0 var(--accent,#5b9dff);animation:lj-pulse 1.6s infinite}
#live-jobs .lj-badge.idle .lj-dot{background:var(--muted,#8a8a90);animation:none}
@keyframes lj-pulse{0%{box-shadow:0 0 0 0 rgba(91,157,255,.5)}70%{box-shadow:0 0 0 6px rgba(91,157,255,0)}100%{box-shadow:0 0 0 0 rgba(91,157,255,0)}}
#live-jobs .lj-list{pointer-events:auto;display:flex;flex-direction:column;gap:6px}
#live-jobs.collapsed .lj-list{display:none}
#live-jobs .lj-job{background:var(--panel,rgba(30,30,34,.94));border:1px solid var(--border,rgba(127,127,127,.24));
  border-left:2px solid var(--border,rgba(127,127,127,.24));border-radius:10px;padding:9px 11px;
  box-shadow:0 4px 16px rgba(0,0,0,.26);backdrop-filter:blur(8px);color:var(--text,#e8e8ea);
  transition:opacity .5s ease}
#live-jobs .lj-job.mine{border-left-color:var(--accent,#5b9dff)}
#live-jobs .lj-job.done{border-left-color:#3fbf6f}
#live-jobs .lj-job.failed{border-left-color:#e5534b}
#live-jobs .lj-job.stalled{border-left-color:#d8a24a}
#live-jobs .lj-job.fading{opacity:0}
#live-jobs .lj-top{display:flex;align-items:baseline;gap:6px;margin-bottom:5px}
#live-jobs .lj-icon{flex:none;opacity:.8}
#live-jobs .lj-label{flex:1;font-weight:600;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}
#live-jobs .lj-pct{flex:none;font-variant-numeric:tabular-nums;color:var(--muted,#a2a2a8);font-size:12px}
#live-jobs .lj-track{height:5px;border-radius:3px;background:var(--border,rgba(127,127,127,.22));overflow:hidden}
#live-jobs .lj-fill{height:100%;width:0;border-radius:3px;background:var(--accent,#5b9dff);transition:width .4s ease}
#live-jobs .lj-job.done .lj-fill{background:#3fbf6f}
#live-jobs .lj-job.failed .lj-fill{background:#e5534b}
#live-jobs .lj-track.indet .lj-fill{width:35%;animation:lj-indet 1.3s ease-in-out infinite}
@keyframes lj-indet{0%{margin-left:-35%}100%{margin-left:100%}}
#live-jobs .lj-meta{display:flex;gap:8px;margin-top:5px;color:var(--muted,#9a9aa0);font-size:11.5px;
  white-space:nowrap;overflow:hidden;text-overflow:ellipsis}
#live-jobs .lj-meta .lj-detail{flex:1;overflow:hidden;text-overflow:ellipsis}
#live-jobs .lj-err{color:#e5847e;font-size:11.5px;margin-top:4px;white-space:normal}
@media (max-width:720px){#live-jobs{right:10px;left:10px;width:auto;bottom:96px}}
`;

function ensureDom() {
  if (els) return els;
  const style = document.createElement('style');
  style.id = 'live-jobs-css';
  style.textContent = CSS;
  document.head.appendChild(style);

  const root = document.createElement('div');
  root.id = 'live-jobs';
  root.innerHTML = `<div class="lj-badge idle"><span class="lj-dot"></span><span class="lj-badge-text">Jobs</span></div><div class="lj-list"></div>`;
  document.body.appendChild(root);

  const badge = root.querySelector('.lj-badge');
  badge.addEventListener('click', () => {
    root.classList.toggle('collapsed');
  });

  els = { root, badge, list: root.querySelector('.lj-list') };
  return els;
}

// ---- render ---------------------------------------------------------------

function currentThread() {
  try { return runtime?.state?.live?.chat?.activeId || null; } catch (_) { return null; }
}

function jobHtml(j, mine) {
  const cls = ['lj-job'];
  if (j.status === 'done') cls.push('done');
  else if (j.status === 'failed') cls.push('failed');
  else if (j.stalled) cls.push('stalled');
  if (mine) cls.push('mine');

  const indet = j.status === 'running' && (j.pct == null);
  const pct = j.status === 'done' ? 100 : (j.pct != null ? j.pct : 0);
  const pctLabel = j.status === 'done' ? '✓'
    : j.status === 'failed' ? '✕'
    : (j.pct != null ? `${Math.round(j.pct)}%` : '');

  const bytes = (j.bytesDone != null && j.bytesTotal)
    ? `${fmtBytes(j.bytesDone)} / ${fmtBytes(j.bytesTotal)}`
    : (j.bytesDone != null ? fmtBytes(j.bytesDone) : '');
  const metaBits = [];
  if (j.status === 'running') {
    if (bytes) metaBits.push(bytes);
    const r = fmtRate(j.rate); if (r) metaBits.push(r);
    const e = fmtEta(j.eta); if (e) metaBits.push('ETA ' + e);
    if (j.stalled) metaBits.length = 0, metaBits.push(`no update in ${j.stalled}s`);
  }

  const meta = (metaBits.length || j.detail)
    ? `<div class="lj-meta">${metaBits.map(esc).join(' · ')}${
        j.detail ? `<span class="lj-detail" title="${esc(j.detail)}">${metaBits.length ? ' · ' : ''}${esc(j.detail)}</span>` : ''}</div>`
    : '';
  const err = j.status === 'failed' && j.error ? `<div class="lj-err">${esc(j.error)}</div>` : '';

  return `<div class="${cls.join(' ')}" data-id="${esc(j.id)}">
    <div class="lj-top">
      <span class="lj-icon">${kindIcon(j.kind)}</span>
      <span class="lj-label" title="${esc(j.label)}">${esc(j.label)}</span>
      <span class="lj-pct">${pctLabel}</span>
    </div>
    <div class="lj-track${indet ? ' indet' : ''}"><div class="lj-fill" style="width:${pct}%"></div></div>
    ${meta}${err}
  </div>`;
}

function render() {
  const { root, badge, list } = ensureDom();
  const running = jobs.filter((j) => j.status === 'running');
  const has = jobs.length > 0;

  if (has) {
    emptySince = 0;
    root.classList.add('show');
    root.classList.remove('fade-hide');
  } else if (root.classList.contains('show')) {
    // schedule a hide once truly empty for a moment
    if (!emptySince) emptySince = Date.now();
    if (Date.now() - emptySince >= FADE_AFTER_MS) root.classList.remove('show');
  }

  const n = running.length;
  badge.classList.toggle('idle', n === 0);
  const txt = badge.querySelector('.lj-badge-text');
  txt.textContent = n > 0 ? `${n} running` : (has ? 'Jobs done' : 'Jobs');

  const cur = currentThread();
  list.innerHTML = jobs.map((j) => jobHtml(j, j.thread && cur && j.thread === cur)).join('');
}

// ---- stream ---------------------------------------------------------------

function apply(payload) {
  if (!payload || !Array.isArray(payload.jobs)) return;
  jobs = payload.jobs;
  render();
}

function connect() {
  try {
    es = new EventSource(STREAM, { withCredentials: true });
  } catch (_) { scheduleReconnect(); return; }

  es.onmessage = (e) => {
    backoff = 1000;
    try { apply(JSON.parse(e.data)); } catch (_) { /* keepalive or garbage */ }
  };
  es.onerror = () => {
    try { es.close(); } catch (_) {}
    es = null;
    scheduleReconnect();
  };
}

function scheduleReconnect() {
  // Fallback: one plain GET so a broken SSE still shows current state.
  fetch(FALLBACK_POLL, { credentials: 'same-origin' })
    .then((r) => r.ok ? r.json() : null).then((d) => d && apply(d)).catch(() => {});
  const wait = Math.min(backoff, 15000);
  backoff = Math.min(backoff * 2, 15000);
  setTimeout(connect, wait);
}

// Re-render on SPA navigation so "mine" highlighting tracks the open thread.
function watchThread() {
  let last = currentThread();
  pollTimer = setInterval(() => {
    const c = currentThread();
    if (c !== last) { last = c; if (jobs.length) render(); }
  }, 800);
}

export function initJobs() {
  if (els) return;                 // idempotent
  ensureDom();
  connect();
  watchThread();
}

// Self-boot on import (app.js just needs to import this module once).
if (typeof window !== 'undefined' && typeof document !== 'undefined') {
  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', initJobs, { once: true });
  } else {
    initJobs();
  }
}
